### PR TITLE
Plugin Directory empty state

### PR DESCRIPTION
**Pull request by jklausa**
This turned out to be a bit bigger than I hoped, because when I was adding the empty state and refresh button, I realised it's kinda broken in the list view, and when I was already there I refactored the featured plugin code to not require a passed `site`...

Hopefully it ain't too big though.

Testing:
1. Turn on Network Link Conditioner, with "100% loss" preset.
2. Go to plugins.
3. Verify that it shows an empty state  indicating that it's loading plugins.
4. Let it timeout
5. Verify that it shows an error state.
6. Turn off the Network Link Conditioner
7. Tap on the "Try Again" button
8. Verify that the plugins loaded correctly

### CODE REVIEW

File: WordPress/Classes/Stores/PluginStore.swift [hunk 1]
```diff
@@ -226,6 +209,21 @@ class PluginStore: QueryStore<PluginStoreState, PluginQuery> {
             .filter { shouldFetchDirectory(feed: $0) }
     }
 
+    private func shouldFetchFeatured() -> Bool {
+        let hasFeaturedQuery = activeQueries.contains {
+            if case .featured = $0 {
```
[the remark this answers is not in the document]
Comment (jklausa) [reply WordPress/Classes/Stores/PluginStore.swift]: fixed in a66ae27

File: WordPress/Classes/ViewRelated/Plugins/PluginDirectoryViewModel.swift [hunk 2]
```diff
@@ -12,36 +12,71 @@ class PluginDirectoryViewModel: Observable {
...
+        featuredReceipt = store.query(.featured)
         popularReceipt = store.query(.feed(type: .popular))
         newReceipt = store.query(.feed(type: .newest))
 
         storeReceipt = store.onChange { [weak self] in
             self?.changeDispatcher.dispatch()
         }
+
+        actionReceipt = ActionDispatcher.global.subscribe { [weak self] (action) in
+            self?.refreshRefreshing()
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Plugins/PluginDirectoryViewModel.swift]: This should happen on `store.onChange`. Nothing should probably subscribe to ActionDispatcher unless it's going to process the given Action

**Comment by jklausa** [comment]
@koke I got inspired on a flight and re-done this completely, it now has activityIndicators inline in the cells and it won't reflow the whole screen on updates anymore.

up for another look!

File: WordPress/Classes/ViewRelated/Plugins/CollectionViewContainerRow.swift [hunk 3]
```diff
@@ -42,22 +44,41 @@ struct CollectionViewContainerRow<CollectionViewCellType: UICollectionViewCell,
...
+        self.secondaryTitle = secondaryTitle
+        self.action = action
+        self.noResultsView = noResultsView
+
+        let configurationBlock = { (_: UICollectionViewCell, _: Any) in return }
+        let selectionBlock = { (_: Any) in return }
+
+        helper = CollectionViewHandler(dataSourceItems: [],
+                                       configureCollectionCellBlock: configurationBlock,
+                                       collectionCellSeletectedBlock: selectionBlock)
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Plugins/CollectionViewContainerRow.swift]: You can simplify the empty block syntax:

```swift
        helper = CollectionViewHandler(dataSourceItems: [],
                                       configureCollectionCellBlock: { _, _ in },
                                       collectionCellSeletectedBlock: { _ in })
```

But if this is an option, I'd just make the two blocks optionals and pass `nil`.

File: WordPress/Classes/ViewRelated/Plugins/PluginDirectoryViewModel.swift [hunk 4]
```diff
@@ -34,73 +38,234 @@ class PluginDirectoryViewModel: Observable {
...
+        }
+    }
+
+    func tableViewModel(presenter: PluginPresenter & PluginListPresenter) -> ImmuTable {
+        return ImmuTable(optionalSections: [
+            ImmuTableSection(optionalRows: [
+                installedRow(presenter: presenter),
+                featuredRow(presenter: presenter),
+                popularRow(presenter: presenter),
+                newRow(presenter: presenter),
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Plugins/PluginDirectoryViewModel.swift]: If the four sections are always displayed now, even on initial load, why is it possible for them to be optional?

File: WordPress/Classes/ViewRelated/Plugins/PluginDirectoryViewModel.swift [hunk 5]
```diff
@@ -34,73 +38,234 @@ class PluginDirectoryViewModel: Observable {
         }
     }
 
-    func tableViewModel(presenter: PluginPresenter & PluginListPresenter) -> ImmuTable {
+    func reloadFailed() {
+        allQueries
+            .filter { !isFetching(for: $0) }
+            .filter { !hasResults(for: $0) }
+            .flatMap { refreshAction(for: $0) }
+            .forEach { ActionDispatcher.dispatch($0) }
```
Comment (koke) [inline WordPress/Classes/ViewRelated/Plugins/PluginDirectoryViewModel.swift]: I think the `isFetching` logic belongs in the Store, and IIRC it was already checking for that when processing refresh actions.

**Pull request merged**